src/main/java/org/opensearch/knn/index/query/KNNWeight.java

```diff
@@ -153,6 +153,26 @@ public Map<Integer, Float> searchLeaf(LeafReaderContext context, int k) throws I
         return docIdsToScoreMap;
     }
 
+    /**
+     * For given {@link LeafReaderContext}, this api will return will KNNWeight perform exact search or not
+     * always. This decision is based on two properties, 1) if there are no native engine files in segments,
+     * exact search will always be performed, 2) if number of docs after filter is less than 'k'
+     * @param context
+     * @return
+     * @throws IOException
+     */
+    public boolean isExactSearchPreferred(LeafReaderContext context) throws IOException {
+        final BitSet filterBitSet = getFilteredDocsBitSet(context);
```

Comment:
For each ANN leaf, getting the filter Bitset is now happening twice. Once for this check, and then for getting the actual [filter bitset](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/KNNWeight.java#L129). 

Considering worst case scenario, This goes through [filterWeight.scorer](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/KNNWeight.java#L164) twice and then creating a bitset which involves a [linear loop](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/KNNWeight.java#L184)

How confident are we that this won't impact latencies for filtering cases? 

Can we avoid this duplicate? One way is to do only engine empty check here. and then pass the bitset in [searchLeaf](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/query/KNNWeight.java#L128) and do exact Search  context

Let me know if I am missing something